### DISCUSSION TITLE
Protect source code with pyinstaller

### DISCUSSION THREAD

**Discussion by maxwxzheng** [pyinstaller/pyinstaller#5687] [PyInstaller]
Hi! I use pyinstaller to package our applications. One thing I've been concerned about is reverse-engineering.

From what I read so far, there are tools such as pyinstxtractor and uncompyle6 that make it really easy to reverse-engineer a binary built by pyinstaller. I also understand that it is impossible to completely stop ppl from reverse-engineering the code as long as they have the binary. However, I'd like to make it as hard as possible.

One thing I've been thinking about is to compile all of our source .py files into .so files with cython and somehow make pyinstaller use the .so files to package the application. I'm wondering if this is doable at all.

Another idea is to change the pyinstaller code to compile the .py files into .so files while packaging the app. I'm also not sure if this would be easy to do.

Any thoughts on this? Also are there any other ways to protect source code with pyinstaller?

**Comment by rokm** [comment]
> One thing I've been thinking about is to compile all of our source .py files into .so files with cython and somehow make pyinstaller use the .so files to package the application. I'm wondering if this is doable at all.

It probably is (there are python packages out there that come with cython'ed extensions, and they can be made work with pyinstaller), but you'll need to watch out for two things:
* entry-point script still needs to be a .py file (although it only needs to call an entry-point function from you .so file)
* you'll need to sort out the first-order dependencies of your program/.so library (i.e., other imported python packages) yourself via `hiddenimports`, as PyInstaller's automatic analysis will fail to pick them up

**Reply by maxwxzheng** [reply]
Thanks @rokm! Do you know if there is any easy way to collect all the "other imported python packages"? Does it only need the external packages that we depend on, or does it also need to include all the builtin python libraries that we use?

**Reply by rokm** [reply]
> Do you know if there is any easy way to collect all the "other imported python packages"?

Perhaps you can freeze non-compiled version of your application and then check the resulting import graph in `build/<name>/xref-<name>.html`? But personally, I'd just manually keep expanding `hiddenimports` until there are no missing-module errors left...

> Does it only need the external packages that we depend on, or does it also need to include all the builtin python libraries that we use?

You'll also need to include imported modules from python library (at least the ones [that aren't collected by PyInstaller into `baselib.zip` as core dependencies](https://github.com/pyinstaller/pyinstaller/blob/b2e062c47b7990c708c6524137509bfd4d00fa90/PyInstaller/compat.py#L610)).

Take a look at our [pydantic hook](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/master/src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-pydantic.py) for an example, as recent wheels of that package come with all its modules compiled into extensions.

**Reply by maxwxzheng** [reply]
Gotcha. Thanks for the suggestions. @rokm 
What do you think about this:
> Another idea is to change the pyinstaller code to compile the .py files into .so files while packaging the app. I'm also not sure if this would be easy to do.

Do you think this is doable? It might be a nice addition to pyinstaller overall.

**Reply by bwoodsend** [reply]
Uhmm, there is no way in hell I am implementing that. We'd receive a torrent of Windows users who can't install MSVC, yet more macOS and Windows users having trouble with anti-viruses and no end of Linux users who are hopelessly naive about glibc incompatibilities and the need to compile any binary on old docker images.

**Reply by rokm** [reply]
I'm not really familiar with `cython` enough to tell whether this could be done automatically or not. But I imagine there are always corner cases that need manual handling and tweaking... 

So if someone wants to cythonize their code for whatever reason, it's probably best if they do it themselves, and fully test the result before trying to package it with PyInstaller. If this was all merged into a single pipeline, it'll make it even harder to determine where any potential issues are coming from.

**Comment by bwoodsend** [comment]
You might be better off with nuitka. It turns everything into C++ then compiles it so everything will be binary. Failing that, you can [mangle variable names](http://liftoff.github.io/pyminifier/#special-sauce) or use [pyarmour](https://pypi.org/project/pyarmor/).

**Reply by maxwxzheng** [reply]
Thanks @bwoodsend I've been playing with Nuitka as well. I'm having trouble getting it to work (we use django and pipenv, which might bring another level of complexity).
I'll check out mangle variable names and pyarmour.

**Comment by AmuCh1** [comment]
I am also stuck on this part. i have entry-point script myapp.py and myapp.so file in the same directory. with my entrypoint script having simply few lines 

from myapp import main

if __name__ == "__main__":
    main()

But when i do  pyinstaller myapp.spec --dist ../distC/ --noconfirm
it says 
 File "/global/freeware/Linux/RHEL7/python-3.6.7/lib/python3.6/site-packages/PyInstaller/archive/writers.py", line 373, in add
    code = get_code_object(nm, pathnm)
  File "/global/freeware/Linux/RHEL7/python-3.6.7/lib/python3.6/site-packages/PyInstaller/building/utils.py", line 621, in get_code_object
    raise ValueError("Module file %s is missing" % filename)
ValueError: Module file <mypath>/myapp.py is missing

I added     binaries=[('myapp.so','.')],  in my spec file. 

can you please help. thanks,
Amu-

**Reply by rokm** [reply]
`ValueError: Module file /myapp.py is missing` 

This looks like invalid path to the `myapp.py` script in the spec file (note the leading `/`)...

**Reply by rokm** [reply]
Or rather, it could be that the problem is that both the entry-point script and the compiled module are named `myapp`, and end up shadowing each other. Try renaming the entry-point script.
